File: zoo/src/main/scala/com/intel/analytics/zoo/models/recommendation/NeuralCF.scala
/*
 * Copyright 2018 Analytics Zoo Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.intel.analytics.zoo.models.recommendation

import com.intel.analytics.bigdl.nn.Graph.ModuleNode
import com.intel.analytics.bigdl.nn.abstractnn.AbstractModule
import com.intel.analytics.bigdl.tensor.Tensor
import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
import com.intel.analytics.bigdl.utils.Shape
import com.intel.analytics.zoo.models.common.ZooModel
import com.intel.analytics.zoo.pipeline.api.keras.layers._
import com.intel.analytics.zoo.pipeline.api.keras.models.Model

import scala.reflect.ClassTag

/**
 * The neural collaborative filtering model used for recommendation.
 * @param userCount    The number of users. Positive integer.
 * @param itemCount    The number of items. Positive integer.
 * @param numClasses   The number of classes. Positive integer.
 * @param userEmbed    Units of user embedding. Positive integer. Default is 20.
 * @param itemEmbed    Units of item embedding. Positive integer. Default is 20.
 * @param hiddenLayers Units hiddenLayers for MLP. Array of positive integers.
 *                     Default is Array(40, 20, 10).
 * @param includeMF    Whether to include Matrix Factorization. Boolean. Default is true.
 * @param mfEmbed      Units of matrix factorization embedding. Positive integer.
 *                     Default is 20.
 * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now.
 */

class NeuralCF[T: ClassTag](
    val userCount: Int,
    val itemCount: Int,
    val numClasses: Int,
    val userEmbed: Int = 20,
    val itemEmbed: Int = 20,
    val hiddenLayers: Array[Int] = Array(40, 20, 10),
    val includeMF: Boolean = true,
    val mfEmbed: Int = 20)(implicit ev: TensorNumeric[T])
  extends Recommender[T] {

  override def buildModel(): AbstractModule[Tensor[T], Tensor[T], T] = {
    val input = Input[T](inputShape = Shape(2))

    val userSelect = Select[T](1, 0).inputs(input)
    val itemSelect = Select[T](1, 1).inputs(input)

    val userFlat = Flatten().inputs(userSelect)
    val itemFlat = Flatten().inputs(itemSelect)

    val mlpUserTable = Embedding[T](userCount + 1, userEmbed, init = "normal")
    val mlpItemTable = Embedding[T](itemCount + 1, itemEmbed, init = "normal")

    val mlpUserLatent: ModuleNode[T] = Flatten().inputs(mlpUserTable.inputs(userFlat))
    val mlpItemLatent: ModuleNode[T] = Flatten().inputs(mlpItemTable.inputs(itemFlat))

    val mlpEmbeddedLayer = Merge.merge[T](List(mlpUserLatent, mlpItemLatent), "concat", 1)

    val linear1 = Dense[T](hiddenLayers(0), activation = "relu").inputs(mlpEmbeddedLayer)
    var mlpLinear = linear1
    for (i <- 1 to hiddenLayers.length - 1) {
      val linearMid = Dense(hiddenLayers(i), activation = "relu").inputs(mlpLinear)
      mlpLinear = linearMid
    }

    val linearLast = if (includeMF) {
      require(mfEmbed > 0, s"please provide meaningful number of embedding units")

      val mfUserTable = Embedding[T](userCount + 1, mfEmbed, init = "normal")
      val mfItemTable = Embedding[T](itemCount + 1, mfEmbed, init = "normal")

      val mfUserLatent: ModuleNode[T] = Flatten().inputs(mfUserTable.inputs(userFlat))
      val mfItemLatent: ModuleNode[T] = Flatten().inputs(mfItemTable.inputs(itemFlat))

      val mfEmbeddedLayer = Merge.merge[T](List(mfUserLatent, mfItemLatent), "mul", 1)

      val concatedModel = Merge.merge[T](List(mlpLinear, mfEmbeddedLayer), "concat", 1)

      Dense(numClasses, activation = "softmax").inputs(concatedModel)
    }
    else {
      Dense(numClasses, activation = "softmax").inputs(mlpLinear)
    }

    val model = Model[T](input, linearLast)

    model.asInstanceOf[AbstractModule[Tensor[T], Tensor[T], T]]
  }
}

object NeuralCF {
  /**
   * The factory method to create a NeuralCF instance.
   */
  def apply[@specialized(Float, Double) T: ClassTag](
      userCount: Int,
      itemCount: Int,
      numClasses: Int,
      userEmbed: Int = 20,
      itemEmbed: Int = 20,
      hiddenLayers: Array[Int] = Array(40, 20, 10),
      includeMF: Boolean = true,
      mfEmbed: Int = 20)(implicit ev: TensorNumeric[T]): NeuralCF[T] = {
    new NeuralCF[T](userCount, itemCount, numClasses, userEmbed,
      itemEmbed, hiddenLayers, includeMF, mfEmbed).build()
  }

  /**
   * Load an existing NeuralCF model (with weights).
   *
   * @param path       The path for the pre-defined model.
   *                   Local file system, HDFS and Amazon S3 are supported.
   *                   HDFS path should be like "hdfs://[host]:[port]/xxx".
   *                   Amazon S3 path should be like "s3a://bucket/xxx".
   * @param weightPath The path for pre-trained weights if any. Default is null.
   * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now.
   */
  def loadModel[T: ClassTag](
      path: String,
      weightPath: String = null)(implicit ev: TensorNumeric[T]): NeuralCF[T] = {
    ZooModel.loadModel(path, weightPath).asInstanceOf[NeuralCF[T]]
  }
}

File: pyzoo/zoo/models/recommendation/neuralcf.py
#
# Copyright 2018 Analytics Zoo Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import sys

from zoo.models.common import KerasZooModel
from zoo.models.recommendation import Recommender
from zoo.pipeline.api.keras.layers import *
from zoo.pipeline.api.keras.models import *
from zoo.common.utils import callZooFunc

if sys.version >= '3':
    long = int
    unicode = str


class NeuralCF(Recommender):
    """
    The neural collaborative filtering model used for recommendation.

    # Arguments
    user_count: The number of users. Positive int.
    item_count: The number of classes. Positive int.
    class_num: The number of classes. Positive int.
    user_embed: Units of user embedding. Positive int. Default is 20.
    item_embed: itemEmbed Units of item embedding. Positive int. Default is 20.
    hidden_layers: Units of hidden layers for MLP. Tuple of positive int. Default is (40, 20, 10).
    include_mf: Whether to include Matrix Factorization. Boolean. Default is True.
    mf_embed: Units of matrix factorization embedding. Positive int. Default is 20.
    """

    def __init__(self, user_count, item_count, class_num, user_embed=20,
                 item_embed=20, hidden_layers=[40, 20, 10], include_mf=True,
                 mf_embed=20, bigdl_type="float"):
        self.user_count = int(user_count)
        self.item_count = int(item_count)
        self.class_num = int(class_num)
        self.user_embed = int(user_embed)
        self.item_embed = int(item_embed)
        self.hidden_layers = [int(unit) for unit in hidden_layers]
        self.include_mf = include_mf
        self.mf_embed = int(mf_embed)
        self.bigdl_type = bigdl_type
        self.model = self.build_model()
        super(NeuralCF, self).__init__(None, self.bigdl_type,
                                       self.user_count,
                                       self.item_count,
                                       self.class_num,
                                       self.user_embed,
                                       self.item_embed,
                                       self.hidden_layers,
                                       self.include_mf,
                                       self.mf_embed,
                                       self.model)

    def build_model(self):
        input = Input(shape=(2,))
        user_flat = Flatten()(Select(1, 0)(input))
        item_flat = Flatten()(Select(1, 1)(input))
        mlp_user_embed = Embedding(self.user_count + 1, self.user_embed, init="uniform")(user_flat)
        mlp_item_embed = Embedding(self.item_count + 1, self.item_embed, init="uniform")(item_flat)
        mlp_user_flat = Flatten()(mlp_user_embed)
        mlp_item_flat = Flatten()(mlp_item_embed)
        mlp_latent = merge(inputs=[mlp_user_flat, mlp_item_flat], mode="concat")
        linear1 = Dense(self.hidden_layers[0], activation="relu")(mlp_latent)
        mlp_linear = linear1
        for ilayer in range(1, len(self.hidden_layers)):
            linear_mid = Dense(self.hidden_layers[ilayer], activation="relu")(mlp_linear)
            mlp_linear = linear_mid

        if (self.include_mf):
            assert (self.mf_embed > 0)
            mf_user_embed = Embedding(self.user_count + 1, self.mf_embed, init="uniform")(user_flat)
            mf_item_embed = Embedding(self.item_count + 1, self.mf_embed, init="uniform")(item_flat)
            mf_user_flatten = Flatten()(mf_user_embed)
            mf_item_flatten = Flatten()(mf_item_embed)
            mf_latent = merge(inputs=[mf_user_flatten, mf_item_flatten], mode="mul")
            concated_model = merge(inputs=[mlp_linear, mf_latent], mode="concat")
            linear_last = Dense(self.class_num, activation="softmax")(concated_model)
        else:
            linear_last = Dense(self.class_num, activation="softmax")(mlp_linear)
        model = Model(input, linear_last)
        return model

    @staticmethod
    def load_model(path, weight_path=None, bigdl_type="float"):
        """
        Load an existing NeuralCF model (with weights).

        # Arguments
        path: The path for the pre-defined model.
              Local file system, HDFS and Amazon S3 are supported.
              HDFS path should be like 'hdfs://[host]:[port]/xxx'.
              Amazon S3 path should be like 's3a://bucket/xxx'.
        weight_path: The path for pre-trained weights if any. Default is None.
        """
        jmodel = callZooFunc(bigdl_type, "loadNeuralCF", path, weight_path)
        model = KerasZooModel._do_load(jmodel, bigdl_type)
        model.__class__ = NeuralCF
        return model
